File: frontend/src/services/opportunity.ts
/**
 * 商机服务
 */
import apiClient from './api'
import { API_ENDPOINTS } from '../config/api'
import { PaginatedResponse } from '../types/common'

export interface CollaborativeMember {
  id: number
  opportunity_id: number
  member_id: number
  role?: string
  description?: string
  created_at: string
  // 关联信息
  member_name?: string
  member_username?: string
}

export interface Opportunity {
  id: number
  opportunity_no: string
  lead_id: number
  task_id: number
  team_leader_id: number
  customer_unit: string
  required_products?: string
  description?: string
  expected_amount?: string
  status: string
  lost_reason?: string
  won_amount?: string
  created_at: string
  updated_at: string
  status_changed_at?: string
  collaborative_members?: CollaborativeMember[]
  // 关联信息
  lead_customer_name?: string
  lead_requirement_direction?: string
  task_name?: string
  task_sales_unit?: string
  team_leader_name?: string
  team_leader_username?: string
}

export interface OpportunityCreate {
  opportunity_no: string
  lead_id: number
  task_id: number
  team_leader_id: number
  customer_unit: string
  required_products?: string
  description?: string
  expected_amount?: string
}

export interface OpportunityUpdate {
  customer_unit?: string
  required_products?: string
  description?: string
  expected_amount?: string
  status?: string
  lost_reason?: string
  won_amount?: string
}

export interface CollaborativeMemberCreate {
  member_id: number
  role?: string
  description?: string
}

export interface OpportunityListParams {
  task_id?: number
  status?: string
  member_id?: number
  required_product?: string
  start_date?: string
  end_date?: string
  search?: string
  page?: number
  page_size?: number
}

export const opportunityService = {
  /**
   * 获取商机列表（支持分页和搜索）
   */
  getOpportunities: async (params?: OpportunityListParams): Promise<PaginatedResponse<Opportunity>> => {
    const queryParams: any = {}
    if (params?.task_id) queryParams.task_id = params.task_id
    if (params?.status) queryParams.status = params.status
    if (params?.member_id) queryParams.member_id = params.member_id
    if (params?.required_product) queryParams.required_product = params.required_product
    if (params?.start_date) queryParams.start_date = params.start_date
    if (params?.end_date) queryParams.end_date = params.end_date
    if (params?.search) queryParams.search = params.search
    if (params?.page) queryParams.page = params.page
    if (params?.page_size) queryParams.page_size = params.page_size
    return apiClient.get(API_ENDPOINTS.OPPORTUNITIES, { params: queryParams })
  },

  /**
   * 获取商机详情
   */
  getOpportunity: async (id: number): Promise<Opportunity> => {
    return apiClient.get(API_ENDPOINTS.OPPORTUNITY_DETAIL(id))
  },

  /**
   * 创建商机
   */
  createOpportunity: async (data: OpportunityCreate): Promise<Opportunity> => {
    return apiClient.post(API_ENDPOINTS.OPPORTUNITIES, data)
  },

  /**
   * 更新商机
   */
  updateOpportunity: async (id: number, data: OpportunityUpdate): Promise<Opportunity> => {
    return apiClient.put(API_ENDPOINTS.OPPORTUNITY_DETAIL(id), data)
  },

  /**
   * 添加协同人员
   */
  addCollaborativeMember: async (id: number, data: CollaborativeMemberCreate): Promise<CollaborativeMember> => {
    return apiClient.post(API_ENDPOINTS.OPPORTUNITY_COLLABORATIVE_MEMBERS(id), data)
  },
}

File: backend/app/schemas/opportunity.py
"""
商机相关的 Pydantic 模式
"""
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
from app.models.opportunity import OpportunityStatus


class CollaborativeMemberBase(BaseModel):
    """协同人员基础模式"""
    member_id: int
    role: Optional[str] = None
    description: Optional[str] = None


class CollaborativeMemberCreate(CollaborativeMemberBase):
    """创建协同人员模式（opportunity_id在URL路径中提供，不需要在请求体中）"""
    pass


class CollaborativeMemberResponse(CollaborativeMemberBase):
    """协同人员响应模式（包含关联信息）"""
    id: int
    opportunity_id: int
    created_at: datetime
    # 成员信息
    member_name: Optional[str] = None
    member_username: Optional[str] = None
    
    @classmethod
    def from_orm_with_relations(cls, member):
        """从ORM对象创建响应，包含关联信息"""
        data = {
            "id": member.id,
            "opportunity_id": member.opportunity_id,
            "member_id": member.member_id,
            "role": member.role,
            "description": member.description,
            "created_at": member.created_at,
        }
        
        # 填充成员信息
        if member.member:
            data["member_name"] = member.member.real_name
            data["member_username"] = member.member.username
        
        return cls(**data)
    
    class Config:
        from_attributes = True


class OpportunityBase(BaseModel):
    """商机基础模式"""
    opportunity_no: str
    lead_id: int  # 关联线索ID（商机必须从线索创建）
    task_id: int
    customer_unit: str
    required_products: str  # 具体产品（必填）
    description: str  # 商机描述（必填）
    expected_amount: Optional[str] = None  # 预计金额（可选）


class OpportunityCreate(OpportunityBase):
    """创建商机模式（从线索转换）"""
    team_leader_id: int


class OpportunityUpdate(BaseModel):
    """更新商机模式"""
    customer_unit: Optional[str] = None
    required_products: Optional[str] = None
    description: Optional[str] = None
    expected_amount: Optional[str] = None
    status: Optional[OpportunityStatus] = None
    lost_reason: Optional[str] = None
    won_amount: Optional[str] = None


class OpportunityInDB(OpportunityBase):
    """数据库中的商机模式"""
    id: int
    lead_id: int
    team_leader_id: int
    status: OpportunityStatus
    expected_amount: Optional[str] = None
    lost_reason: Optional[str] = None
    won_amount: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    status_changed_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


class OpportunityResponse(OpportunityInDB):
    """商机响应模式（包含关联信息）"""
    collaborative_members: List[CollaborativeMemberResponse] = []
    # 线索信息
    lead_id: Optional[int] = None
    lead_customer_name: Optional[str] = None
    lead_requirement_direction: Optional[str] = None
    
    # 任务信息
    task_name: Optional[str] = None
    task_sales_unit: Optional[str] = None
    
    # 组长信息
    team_leader_name: Optional[str] = None
    team_leader_username: Optional[str] = None
    
    @classmethod
    def from_orm_with_relations(cls, opportunity, db=None):
        """从ORM对象创建响应，包含关联信息"""
        from app.utils.product_utils import convert_product_value_to_label
        
        # 转换required_products：将value转换为label
        required_products_display = opportunity.required_products
        if db and opportunity.required_products:
            required_products_display = convert_product_value_to_label(
                opportunity.required_products, db
            )
        
        data = {
            "id": opportunity.id,
            "opportunity_no": opportunity.opportunity_no,
            "lead_id": opportunity.lead_id,
            "task_id": opportunity.task_id,
            "team_leader_id": opportunity.team_leader_id,
            "customer_unit": opportunity.customer_unit,
            "required_products": required_products_display,
            "description": opportunity.description,
            "expected_amount": opportunity.expected_amount,
            "status": opportunity.status,
            "lost_reason": opportunity.lost_reason,
            "won_amount": opportunity.won_amount,
            "created_at": opportunity.created_at,
            "updated_at": opportunity.updated_at,
            "status_changed_at": opportunity.status_changed_at,
        }
        
        # 填充线索信息
        if opportunity.lead:
            data["lead_id"] = opportunity.lead.id
            data["lead_customer_name"] = opportunity.lead.customer_name
            data["lead_requirement_direction"] = opportunity.lead.requirement_direction
        
        # 填充任务信息
        if opportunity.task:
            data["task_name"] = opportunity.task.task_name
            data["task_sales_unit"] = opportunity.task.sales_unit
        
        # 填充组长信息
        if opportunity.team_leader:
            data["team_leader_name"] = opportunity.team_leader.real_name
            data["team_leader_username"] = opportunity.team_leader.username
        
        # 处理协同人员
        collaborative_members = []
        if hasattr(opportunity, 'collaborative_members') and opportunity.collaborative_members:
            for member in opportunity.collaborative_members:
                collaborative_members.append(CollaborativeMemberResponse.from_orm_with_relations(member))
        data["collaborative_members"] = collaborative_members
        
        return cls(**data)

